src/platform/openiotsdk/PlatformManagerImpl.cpp

```diff
@@ -152,74 +180,83 @@ void PlatformManagerImpl::HandleTimerEvent(void)
     }
 }
 
-void PlatformManagerImpl::RunEventLoopInternal()
+void PlatformManagerImpl::_RunEventLoop()
 {
     uint32_t flags = 0;
 
-    while (true)
+    if (!mInitialized)
     {
-#if CHIP_SYSTEM_CONFIG_USE_IOT_SOCKET
-        if (SystemLayerImpl().WaitForEvents() == CHIP_NO_ERROR)
+        ChipLogError(DeviceLayer, "_PostEvent: stack not initialized");
+        return;
+    }
+
+    {
+        ScopedLock lock(mEventTaskMutex);
+
+        bool expectedValue = false;
+        if (!mRunEventLoop.compare_exchange_strong(expectedValue /* expected */, true /* desired */))
         {
-            LockChipStack();
-            SystemLayerImpl().HandleEvents();
-            UnlockChipStack();
+            ChipLogError(DeviceLayer, "Error trying to run the event loop while it is already running");
+            return;
         }
 
-        flags = osEventFlagsGet(mPlatformFlags);
-#else
-        flags = osEventFlagsWait(mPlatformFlags, kPostEventFlag | kTimerEventFlag | kTaskStopEventFlag,
-                                 osFlagsWaitAny | osFlagsNoClear, ms2tick(1000));
-        // in case of error we still need to know the value of flags we're not waiting for
-        if (flags & osFlagsError)
+        // Look if a task ID has already been assigned or not.
+        // If not, it means we run in the thread that called RunEventLoop
+        if (!mEventTask)
         {
-            flags = osEventFlagsGet(mPlatformFlags);
+            ChipLogDetail(DeviceLayer, "Run CHIP event loop on external thread");
+            mEventTask = osThreadGetId();
+        }
+        else
+        {
+            osEventFlagsSet(mPlatformFlags, kTaskHasEventLoopRunFlag);
         }
+    }
+
+    LockChipStack();
+
+    while (mRunEventLoop.load())
+    {
+        UnlockChipStack();
```

Comment:
Why are we Unlocking here ? Doesn't it means that the loop will be exited while the ChipStack is locked ? Why not unlock at the end of the loop ?

Reply:
The idea is to unlock the stack only for waiting for the event, Events service should be under the locked stack. We unlock when we leave the loop.